### ISSUE TITLE
 Retrieve pagination information

### ISSUE THREAD

**Issue by schickling** [prisma/prisma#4990]
### Problem

Prisma client currently does not provide the ability to retrieve pagination information about a certain query, which prevents developers from accounting for a result set’s size in a variety of cases (display paginated result lists, optimizing batch size…)

### Proposed solution

Add support for a `$withPageInfo()` to functions returning collections (`findMany` ,1-n relations, …) returning pagination information alongside the results.

The pagination information could be described as:

```ts
type PageInfo<T> = {
  data: T[] // The query result
  hasNext: boolean // Indication that the cursor can move to the next page
  hasPrev: boolean // Indication that the cursor can move to the previous page
}
```

Example usage:

```ts
const samsPostsWithPageInfo: PageInfo<Post> = await prisma.users
  .findOne('sams-id')
  .posts({ first: 50 })
  .$withPageInfo()
console.log(samsPostsWithPageInfo.hasNext, samsPostsWithPageInfo.hasPrev, samsPostsWithPageInfo.data
```

**Comment by johannesschobel** [comment]
i would also need this feature. Currently, i have a custom solution, but a default prisma solution would be preferred ;)

**Comment by afulkersonApollo** [comment]
Bump? What's the possibility of getting this added? Seems easy and it's a feature that Sequelize has had for a while.

**Comment by Karl-EdwardFPJeanMehu** [comment]
I had initially solved this by using only one query, `findMany` without `cursor` or `take`. After grabbing the length of the request, I did the cursor and take and etc myself. Upon discovering `$transaction` I simply did something like the following instead:

```js
let [ result, count ] = await prisma.$transaction([
   prisma.entity.findMany(query),
   prisma.entity.count({where})
])
```

**Comment by afulkersonApollo** [comment]
> I had initially solved this by using only one query, `findMany` without `cursor` or `take`. After grabbing the length of the request, I did the cursor and take and etc myself. Upon discovering `$transaction` I simply did something like the following instead:
> 
> ```js
> let [ result, count ] = await prisma.$transaction([
>    prisma.entity.findMany(query),
>    prisma.entity.count({where})
> ])
> ```

It's still two queries though.

**Comment by Karl-EdwardFPJeanMehu** [comment]
Of course it is @afulkersonApollo but less painful to use rather than juggling with each one and implementing cursor and take logic yourself. I find it a lot more simple and flexible (if both results are mandatory)

**Comment by afulkersonApollo** [comment]
> Of course it is @afulkersonApollo but less painful to use rather than juggling with each one and implementing cursor and take logic yourself. I find it a lot more simple and flexible (if both results are mandatory)

Thanks for the suggestion :)

**Comment by johannesschobel** [comment]
> > I had initially solved this by using only one query, `findMany` without `cursor` or `take`. After grabbing the length of the request, I did the cursor and take and etc myself. Upon discovering `$transaction` I simply did something like the following instead:
> > ```js
> > let [ result, count ] = await prisma.$transaction([
> >    prisma.entity.findMany(query),
> >    prisma.entity.count({where})
> > ])
> > ```
> 
> It's still two queries though.

actually i am not sure, if you can even make this in one single query?
i mean you can make a nice wrapper function, but internally its still 2 queries!? 🤔

**Comment by Karl-EdwardFPJeanMehu** [comment]
@johannesschobel. No, no. Allow me to clarify. My initial solution did in fact use a single query using `findMany` method the query minus the `cursor`, `orderBy', and `take` fields. Here is basically what I did:

```js
const query = {
  where,
  include: {
    product: true,
  }
}

 if (!!before && !!after) throw new Error("Ambiguous query. You cannot use 'before' and 'after' together");
 if (!!first && !!last) throw new Error("Ambiguous query. You cannot use 'first' and 'last' together");

// First and only query
const results = await prisma.entity.findMany(query);
const count = results.length;

// handles `cursor`, `take` and Returns `hasPreviousPage` and `hasNextPage` for pagination 
// Notice `before`, `after`, `first`, `last`
const {hasPreviousPage, hasNextPage, edges}  = getPagination(resultEdges, before, after, first, last);

// Returns pageInfo (including start and end cursors)
const pageInfo = await getPageInfo(resultEdges, hasPreviousPage, hasNextPage)

// returns the final cursor connection spec structure
return getGraphQLConnection(edges, pageInfo, count)

```

I only meant to say that this:

```js
let [ result, count ] = await prisma.$transaction([
   prisma.entity.findMany(query),
   prisma.entity.count({where})
])
``` 
felt less hacky and I did required less code compared to the above. One is more convenient but requires two queries and the other requires more code but only one query.

**Comment by johannesschobel** [comment]
yeah, of course.. but your solution does not work, if you're using paginated result sets with limit and skip, right? if you want to use limit, you need to make a dedicated second query to get the overall count (without the limit and skip)

**Comment by Karl-EdwardFPJeanMehu** [comment]
@johannesschobel. Actually, no. Since I'm leaving out limit, skip and etc from the query, it does not affect `count`. All of those options are handled among the helper functions I created.

**Comment by old-zoomer** [comment]
@pantharshit00 This looks like basic feature that a lot of pagination use cases need. Are there any updates on this?

**Comment by JejeDurden** [comment]
Much needed feature to build a public API with Prisma. Making several queries to correctly format the answer is costly and time consuming. Please make it happen! 🙏🏻

**Comment by cristiancundari** [comment]
This feature would be awesome. It seems like a basic function to have indeed

**Comment by NCSmit** [comment]
+1 this would be a very helpful feature

**Comment by ChristianVaughn** [comment]
This would be nice to have something similar to Sequelize which has a findAndCountAll() function. but even that is just doing 2 queries under the hood i believe. Most useful thing in my opinion would be allowing prisma.model.count() to take in an includes section so you can at least do counts on complex queries that interact with relation models

**Comment by toritsejuFO** [comment]
Still weird this feature doesn't exist natively in the core library

**Comment by BonjourYY** [comment]
Is there any latest progress on this issue? I want to know why this feature hasn't been added.

**Comment by igalklebanov** [comment]
Hey 👋 

Closing this as it's dusty. Feel free to submit a new issue with modern APIs.
